### PR TITLE
Corelli powder calibration: Allow for fixed L1 in the creation of a calibration

### LINKED ISSUE #30302: Corelli powder calibration: Allow for fixed L1 in the creation of a calibration
For details, refer to [#SCD176](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/176)

### PR DESCRIPTION
**Pull request by jmborr**
**Description of work.**

- [x] Implement properties for a fixed source distance
- [x] modify instrument with the fixed distance, and save to table 
- [x] adapt existing tests
- [x] test passing option to fixed source to sample distance. Test of the location of the source in the instrument and also the table of adjustments
- [x] test `validateinputs` when both `FixedSource` and `AdjustSource` are `True` or `False`
- [x] manual test for the reviewer
- [x] update documentation
- [x] release notes unnecessary as the algorithm has not yet being released

**To test:**  
Download [CORELLI_124036_banks42_87.nxs.zip](https://www.dropbox.com/s/wow3qedcstlpf9m/CORELLI_124036_banks42_87.nxs.zip?dl=0) to `/tmp/` and unzip, then run the following script

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from corelli.calibration.powder import load_and_rebin


Load(Filename='/tmp/CORELLI_124036_banks42_87.nxs.',
         BankName='bank42,bank87',
         OutputWorkspace='sample',
         LoadLogs=False)

CorelliPowderCalibrationCreate(InputWorkspace='sample',
                                                    SourceToSampleDistance=19.997,
                                                    OutputWorkspacesPrefix='sample_',
                                                    TofBinning=[3000,-0.001,16660],
                                                    PeakPositions=[1.3143, 1.3854,1.6967, 1.8587, 2.0781, 2.3995, 2.9388, 4.1561],
                                                    ComponentList='bank42/sixteenpack,bank87/sixteenpack')
```

Verify that the first row in table `sample_adjustments` contains the moderator Z-coordinate at `-19.997`

References #30302

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by clayton-tom — COMMENTED** [review]
I'm getting a `vector subscript out of range` error during the running of ModeratorTzero when running this on windows. Also, just as a small note, in the test script `SourceSampleDistance` needs to be `SourceToSampleDistance` to match the parameter name in the code

**Comment by jmborr** [comment]
> I'm getting a `vector subscript out of range` error during the running of ModeratorTzero when running this on windows. Also, just as a small note, in the test script `SourceSampleDistance` needs to be `SourceToSampleDistance` to match the parameter name in the code

I am not getting the `vector subscript out of range` error:  
Capture


Do you get the same error if you only run `ModeratorTzero` on the loaded file, like this?:

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
Load(Filename='/tmp/CORELLI_124036_banks42_87.nxs.', BankName='bank42,bank87', OutputWorkspace='sample', LoadLogs=False)
ModeratorTzero(InputWorkspace='sample', OutputWorkspace='sample_2', EMode='Elastic')
```
I ran this code snippet and I didn't get any `vector subscript out of range` error

**Comment by clayton-tom** [comment]
I didn't get an error running that snippet, and upon a closer look it seems ModeratorTZero is giving a soft crash every other time or so, while the vector error occurs during PDCalibration as called on L198. It seems slightly fishy as the estimate for the operation is ~58h 
image

**Comment by jmborr** [comment]
I assume you're testing this branch by building it in your local machine, but just to be sure, have you built the branch _after_ rebasing onto an updated `master`?

**Comment by clayton-tom** [comment]
@jmborr I have, but had previously built in Debug mode, I'll try it in release to be sure

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py [hunk 1]
```diff
@@ -30,16 +31,39 @@ def test_exec(self):
...
                                 RelativePosition=False)
+
+        # Both FixSource=True, AdjustSource=True can't be True
+        try:
+            CorelliPowderCalibrationCreate(
+                InputWorkspace='test_workspace', OutputWorkspacesPrefix='cal_', TubeDatabaseDir='/tmp',
+                TofBinning=[300, 1.0, 16666.7], PeakPositions=spacings_reference, FixSource=True, AdjustSource=True,
+                ComponentList='bank1', ComponentMaxTranslation=0.2, ComponentMaxRotation=10)
+        except RuntimeError as error:
+            assert 'Some invalid Properties found' in str(error)
```
Comment (clayton-tom) [inline Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py:42]: Might be worth including the specific property in these error message checks, to make the tests more specific

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py [hunk 2]
```diff
@@ -30,16 +31,39 @@ def test_exec(self):
...
+
+        # Both FixSource=True, AdjustSource=True can't be False
+        try:
+            CorelliPowderCalibrationCreate(
+                InputWorkspace='test_workspace', OutputWorkspacesPrefix='cal_', TubeDatabaseDir='/tmp',
+                TofBinning=[300, 1.0, 16666.7], PeakPositions=spacings_reference, FixSource=False, AdjustSource=False,
+                ComponentList='bank1', ComponentMaxTranslation=0.2, ComponentMaxRotation=10)
+        except RuntimeError as error:
+            assert 'Some invalid Properties found' in str(error)
```
Comment (clayton-tom) [inline Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py:51]: As above

**Review by clayton-tom — CHANGES_REQUESTED** [review]
The algorithm now runs successfully for me and produces the correct value in `sample_adjustments`. Let me know what you think on the test code comments 👍

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py (hunk 1 above)
Comment (jmborr) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py:42]: As you can see [here](https://github.com/mantidproject/mantid/blob/master/Framework/API/src/Algorithm.cpp#L644) the exception message doesn't referent the failing property so I can't extract that information

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py (hunk 2 above)
Comment (jmborr) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py:51]: same reply as above, too :smile:

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py (hunk 1 above)
Comment (clayton-tom) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/CorelliPowderCalibrationCreateTest.py:42]: Understood, sorry, I'd misremembered that output

**Review by clayton-tom — APPROVED** [review]
CI failure unrelated, happy to approve 👍

**Pull request merged**